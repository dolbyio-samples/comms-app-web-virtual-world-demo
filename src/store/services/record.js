import VoxeetSDK from "@voxeet/voxeet-web-sdk";

export const startRecording = () => {
  return VoxeetSDK.recording
    .start()
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};

export const stopRecording = () => {
  return VoxeetSDK.recording
    .stop()
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};
