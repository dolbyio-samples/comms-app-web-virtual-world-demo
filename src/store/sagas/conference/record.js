import { call, select, put } from "redux-saga/effects";

import { startRecording, stopRecording } from "../../services/record";
import { getEntity } from "../../reducers/entities/selectors";
import { getCurrentConferenceID } from "../../reducers/application/selectors";
import { updateConference } from "../../effects/conference";
import { displayMessage } from "./message";

export const getConference = (state, id) => getEntity(state, "conferences", id);

export function* stopRecordingConference() {
  const { error } = yield call(stopRecording);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not stop recording event",
    });

    return;
  }

  const currentConferenceID = yield select(getCurrentConferenceID);
  const conference = yield select(getConference, currentConferenceID);

  yield put(
    updateConference({
      id: conference.id,
      entityType: "conferences",
      data: { recording: false },
    })
  );
}

export function* recordConference() {
  const { error } = yield call(startRecording);
  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not start recording event",
    });

    return;
  }

  const currentConferenceID = yield select(getCurrentConferenceID);
  const conference = yield select(getConference, currentConferenceID);

  yield put(
    updateConference({
      id: conference.id,
      entityType: "conferences",
      data: { recording: true },
    })
  );
}
