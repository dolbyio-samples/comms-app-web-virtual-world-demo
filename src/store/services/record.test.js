import { describe, it, expect, vi } from "vitest";
import { startRecording, stopRecording } from "./record";

const mockStart = vi.fn();
const mockStop = vi.fn();

vi.mock("@voxeet/voxeet-web-sdk", () => ({
  default: {
    get recording() {
      return { start: mockStart, stop: mockStop };
    },
  },
}));

describe("record services test suite", () => {
  it("should succeed", async () => {
    // arrange
    mockStart.mockResolvedValue({ response: {} });

    // act
    const { response, error } = await startRecording();

    // assert
    expect(mockStart).toHaveBeenCalled();
    expect(response).toMatchSnapshot();
    expect(error).toBe(undefined);
  });

  it("should succeed", async () => {
    // arrange
    mockStop.mockResolvedValue({ response: {} });

    // act
    const { response, error } = await stopRecording();

    // assert
    expect(mockStop).toHaveBeenCalled();
    expect(response).toMatchSnapshot();
    expect(error).toBe(undefined);
  });

  it("should fail and return  an error", async () => {
    // arrange
    mockStart.mockRejectedValueOnce(new Error("error"));

    //  act
    const { response, error } = await startRecording();

    // assert
    expect(mockStart).toHaveBeenCalled();
    expect(response).toBe(undefined);
    expect(error.message).toBe("error");
  });

  it("should fail and return an error", async () => {
    // arrange
    mockStop.mockRejectedValueOnce(new Error("error"));

    // act
    const { response, error } = await stopRecording();

    // assert
    expect(mockStop).toHaveBeenCalled();
    expect(response).toBe(undefined);
    expect(error.message).toBe("error");
  });
});
