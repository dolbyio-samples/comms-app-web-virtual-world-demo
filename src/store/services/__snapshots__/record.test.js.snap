// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`record services test suite > should succeed 1`] = `
{
  "response": {},
}
`;

exports[`record services test suite > should succeed 2`] = `
{
  "response": {},
}
`;
